Draft new website homepage, forked from root README.md.

ScottTodd opened the pull request:
Preview is staged at https://scotttodd.github.io/iree/

Progress on https://github.com/google/iree/issues/5461

The sections introduced here relate to the naming discussion at https://github.com/google/iree/issues/5555

antiagainst commented:
Given this is marked as draft, I'd assume you just wanted to seek high-level comments? A few points that I can think of:

* I think we should further trim down the technology density by using more general terms. Reading a landing page with too many of those can be difficult. :)
* I think we should place more focus on what user needs IREE can help to address. Like what framework/platforms/hardware are (gonna be) supported, what flexibility are provided (e.g., not forcing to take on the full stack), why it is better than existing solutions (e.g., native machine code for faster execution, smaller binary size, etc.). Essentially sell a bit more. ;)
* It would be nice to have some graphs/pictures.  Like using some openly available icon set or whatever? You certainly know better than me. (This one is really nice to have so don't spend too much time on it.)

ScottTodd commented:
> Given this is marked as draft, I'd assume you just wanted to seek high-level comments? A few points that I can think of:
> 
> * I think we should further trim down the technology density by using more general terms. Reading a landing page with too many of those can be difficult. :)
> * I think we should place more focus on what user needs IREE can help to address. Like what framework/platforms/hardware are (gonna be) supported, what flexibility are provided (e.g., not forcing to take on the full stack), why it is better than existing solutions (e.g., native machine code for faster execution, smaller binary size, etc.). Essentially sell a bit more. ;)
> * It would be nice to have some graphs/pictures.  Like using some openly available icon set or whatever? You certainly know better than me. (This one is really nice to have so don't spend too much time on it.)

Yep, looking for high level comments, or offers to help with some of the sections. The "Workflow overview" section is most complete, while the introduction and other sections need more thought.

Agreed on all points. Starting with major features, supported platforms, binary size in different configurations, performance targets/benchmarks, etc. sounds good. I started to distill some of the ideas on https://github.com/google/iree#architecture-and-goals here, but we can go further.

I reused the diagram from our README, but it's due for an update. Kojo also requested diagrams on a different page: https://github.com/google/iree/pull/5485#discussion_r617074753.

ScottTodd commented:
Ready for a real review now. I've incorporated your comments @antiagainst :

* Trimmed technical language
* Added more icons and whitespace
* Discussed features and supported platforms
* Filled in missing sections

antiagainst commented on mkdocs/docs/index.md:
```diff
@@ -1,22 +1,156 @@
+provides flexible deployment solutions for its compiled ML models.
 
-## Overview
+## Introduction
```
We should just remove this "Introduction" header? It's not really helpful and a bit disruptive to the flow here (as IMO the above sentence is really the introduction).

ScottTodd replied:
I liked the extra nesting on the table of contents, but agree about disrupting the flow ...

![image](https://user-images.githubusercontent.com/4010439/117221577-6849f180-adbe-11eb-9323-e24c600b96f6.png)

I'll remove it

![image](https://user-images.githubusercontent.com/4010439/117221617-7f88df00-adbe-11eb-9ca5-922306ace079.png)

antiagainst commented on mkdocs/docs/index.md:
```diff
@@ -1,22 +1,156 @@
-## Components
+- [x] Support for dynamic shapes, flow control, streaming, and other advanced
+model features
+- [x] Ahead-of-time compilation of scheduling and execution logic together
```
I would place the 2nd as the 1st bullet as it's arguably the most distinctive feature we have. It's about the architecture. Then we can talk about great support for models (which is the 1st bullet right now), and the 3rd point I think we can cover about great target support (something along the line of "Support for various platforms and various CPU/GPU architectures). (I know we are talking about that in the next section, but that's expanding with more details.) Then we can talk about performance (power/resource efficiency and binary size, etc.). To me that's from a top-down model-to-deployment perspective WDYT?

ScottTodd replied:
Ordering SGTM. I tried a few variations on wording and settled on

> Optimized for many CPU and GPU architectures

Phrasings like

> Compiler and runtime optimized for many CPU and GPU architectures

> Cross platform runtime with optimizations for many CPU and GPU architectures

> Cross platform compiler and runtime optimized for many CPU and GPU architectures

don't quite work for me, since they conflate a few things. Our runtime is "optimized" in the sense that it does as little as possible - we aim to do as much as possible ahead of time in the compiler. The "runtime" itself also doesn't even need to be used at all :P. The compiler is "optimized" in the sense that the compiler itself runs quickly on the host, but also in the sense that it generates efficient code for targets.

antiagainst commented on mkdocs/docs/index.md:
```diff
@@ -1,22 +1,156 @@
+model features
+- [x] Ahead-of-time compilation of scheduling and execution logic together
+- [x] Low overhead, pipelined execution for efficient power and resource usage
+- [x] Binary size as low as 30KB on embedded systems
```
Some additional points to consider: 

- Strongly layered and decoupled components allowing mix and match to serve different usage scenarios
- Provide flexible deployment choices suiting form factor requirements
- ...
?

ScottTodd replied:
I think these are both similar to the "binary size as low as XXKB" bullet. Not quite sure what wording to use there to encompass this too...

antiagainst commented on mkdocs/docs/index.md:
```diff
@@ -1,22 +1,156 @@
+- [x] Low overhead, pipelined execution for efficient power and resource usage
+- [x] Binary size as low as 30KB on embedded systems
+- [x] Debugging and profiling support
+- [ ] Support for running quantized models (coming soon!)
```
Not sure we need a separate point for quantized models. They are too detailed IMO and don't at the same level of points we are discussing here. (It is more fitting into the "Support for dynamic shapes, flow control, streaming, and other advanced model features" part.)

ScottTodd replied:
SG. I'll remove that bullet. I thought it was worth calling out explicitly, since it's a special set of work that performance-focused users will look for on mobile.

antiagainst commented on mkdocs/docs/index.md:
```diff
@@ -1,22 +1,156 @@
+### Support matrix
 
-<!-- TODO(??): explain what frontends are, introduce options -->
+The IREE compiler tools run on :fontawesome-brands-linux: Linux,
```
We should talk about frameworks first and then targets second? (So more coherent with everywhere else.)

But I'm not exactly sure we need this sentence though. We can clear detailed points in the above. It's sort of redundant. (Redundant information is fine if it covers different aspects and is afar. But here it's in the same place.) :)

ScottTodd replied:
Rewording this a bit...

Trying to be concise while getting the point across that we have good support across frontends+backends

antiagainst commented on mkdocs/docs/index.md:
```diff
@@ -1,22 +1,156 @@
+Model formats
 
-<!-- TODO(??): explain what backends are, introduce options -->
+- [x] TensorFlow
```
How do we order this? Should we have some rule? (Like supported -> unsupported, and then alphabetically?)

Similarly for the other tables.

ScottTodd replied:
I loosely ordered first by supported, unsupported, then second by expected # of users

antiagainst commented on mkdocs/docs/index.md:
```diff
@@ -1,22 +1,156 @@
+wide range of platforms, CPU architectures, and hardware accelerators.
 
-### Backends
+Model formats
```
Model authoring frameworks?

(Formats makes me think about GraphDef, protobuf, JSON, etc...)

ScottTodd replied:
ONNX fits into the same category as GraphDef, protobuf, etc. ;) We don't have a path from ONNX yet though.

In this context, I see TensorFlow Lite as less of an authoring framework and more as a converter tool from TF with an associated runtime.

I can get behind "Model authoring frameworks" for now.

antiagainst commented on mkdocs/docs/index.md:
```diff
@@ -1,22 +1,156 @@
+- [x] Windows
+- [x] macOS, iOS
+- [x] Android
+- [x] Bare metal, including RISC-V
```
This is a mix of different dimensions. What about have some empty lines to separate platforms vs. architectures?

ScottTodd replied:
Good idea. I'll spread it out.

antiagainst commented on mkdocs/docs/index.md:
```diff
@@ -1,22 +1,156 @@
+### Selecting deployment configurations
+
+IREE provides a flexible set of tools for various deployment scenarios - use as
+much or as little of IREE as you like.
```
s/use as much or as little of IREE as you like/only pay the cost for use/ to be more positive. :)

ScottTodd replied:
Thought about that, but didn't want to use "pay for what you use" since people might think we're monetizing / charging for features :P

Will think more on phrasing...

ScottTodd replied:
Reworded:

> IREE provides a flexible set of tools for various deployment scenarios.
Fully featured environments can use IREE for dynamic model deployments taking
advantage of multi-threaded hardware, while embedded systems can bypass IREE's
runtime entirely or interface with custom accelerators.

antiagainst commented on mkdocs/docs/index.md:
```diff
@@ -1,22 +1,156 @@
+much or as little of IREE as you like.
+
+* What hardware should the bulk of your model run on? CPU? GPU?
+* What platforms are you targeting? Desktop? Mobile? An embedded system?
```
The 2nd point should be the first one? Logically need to choose the platform first and then see what hardware available.

antiagainst commented on mkdocs/docs/index.md:
```diff
@@ -1,22 +1,156 @@
+### Compiling models
+
+Model compilation is performed ahead-of-time on a _host_ machine for any
+combination of _targets_. The compilation process converts from "ops" used by
```
Change ops to layers/operations so readers are more famililar? (Not sure ops is a known abbr. It might be more known to compiler folks)

antiagainst commented on mkdocs/docs/index.md:
```diff
@@ -1,22 +1,156 @@
+### Running models
+
+IREE offers a low level C API, as well as several specialized sets of
+_bindings_ for running IREE models using other languages.
```
Links to the section?

ScottTodd replied:
Waiting to set up https://squidfunk.github.io/mkdocs-material/setup/setting-up-navigation/#section-index-pages for section linking to work well.

antiagainst commented on mkdocs/docs/index.md:
```diff
@@ -1,22 +1,156 @@
+
+For example, compiling for GPU execution using Vulkan generates SPIR-V kernels
+and Vulkan API calls. For CPU execution, native code with static or dynamic
+linkage and the associated function calls are generated.
```
Links to the section?

antiagainst commented on mkdocs/docs/index.md:
```diff
@@ -1,22 +1,156 @@
+  to support loading different model architectures dynamically?
+
+IREE supports the full set of these configurations using the same underlying
+technology.
```
Links to the section?

antiagainst reviewed (changes requested):
Awesome! Overall structure looks nice. :) I just have a bunch of nits.

antiagainst reviewed (approved):
Awesome. Looks nice to me. Let's land and then iterate.

(One thing that would be nice to improve is that we have quite a few unordered list, which makes the page a bit lengthy. If they can be grouped together with just icons for them, that would be fantastic. But I guess that means more HTML/CSS magic. We can get to it later. :)

ScottTodd commented:
> Awesome. Looks nice to me. Let's land and then iterate.
> 
> (One thing that would be nice to improve is that we have quite a few unordered list, which makes the page a bit lengthy. If they can be grouped together with just icons for them, that would be fantastic. But I guess that means more HTML/CSS magic. We can get to it later. :)

Yep. I was working with https://fontawesome.com/icons?d=gallery&p=2&s=brands&m=free since it was convenient, but we could include logos for more items and put them side by side.
